R/utils.R

```diff
@@ -81,3 +56,49 @@ make_c_call <- function(choices) {
   if (length(private$session_bindings) > 0) lapply(private$session_bindings, function(x) x$destroy())
   invisible(NULL)
 }
+
+
+
+#' Encodes ids to be used in JavaScript and Shiny
+#'
+#' Replaces non-ASCII characters into a format that can be used in HTML,
+#' JavaScript and Shiny.
+#'
+#' Typically, the `%` symbol is used in this format, but we it is not allowed
+#' in this context..
+#' We replace `%` with `__html_XX__` where `XX` is the HTML representation of
+#' the character.
+#'
+#' @param id (`character(1)`) The id string.
+#'
+#' @return Sanitized string that removes special characters and spaces.
+#'
+#' @keywords internal
+js_encode <- function(id) {
+  gsub("%([0-9]{2})", "__html_\\1__", utils::URLencode(as.character(id)))
```

Comment:
I think this is good enough.

However, a more complex alternative would be to convert special characters to utf8 _(the resulting id is more readable)_

``` r
js_encode <- function(x) {
  x_split <- strsplit(gsub("%", " %", utils::URLencode(as.character(x))), " ")[[1]]
  if (length(x_split) == 1) return(x)
  pattern <- "%[0-9]{2}"
  regmatches(x_split, regexpr(pattern, x_split)) <- paste(
    sep = "",
    "__u", 
    as.hexmode(
      sapply(
        utils::URLdecode(regmatches(x_split, regexpr(pattern, x_split))), 
        utf8ToInt, 
        USE.NAMES = FALSE
      )
    ), 
    "__"
  )
  paste(collapse = "", x_split)
}
js_encode(" an id with spaces and %")
#> [1] "__u20__an__u20__id__u20__with__u20__spaces__u20__and__u20____u25__"
js_encode("an_id_with_spaces")
#> [1] "an_id_with_spaces"
```

<sup>Created on 2024-10-23 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
nevermind this. URLencode is already showing utf8 codes

Reply:
I don't think we need a `js_encode`. I think `URLencode` is enough to convert weird ids.
```
utils::URLencode(as.character("iris"))
# [1] "iris"
utils::URLencode(as.character("%iris%"))
# [1] "%25iris%25"
utils::URLencode(as.character("iris.raw"))
# [1] "iris.raw"
utils::URLencode(as.character("raw iris"))
# [1] "raw%20iris"
utils::URLencode(as.character(".raw_data"))
# [1] ".raw_data"
```

Reply:
Quick question - is `"%"` harmful for id?

Reply:
It crashes jquery. You can test in a console with jquery loaded

```js
$("#an-id%20")
```

![image](https://github.com/user-attachments/assets/214284b1-fc9b-461a-ac9a-72b96f3fecb4)

Reply:
> It crashes jquery. You can test in a console with jquery loaded

I think it is ok to use urlEncode if id doesn't work natively with this. I think we should just prepend syntacticly incorrect characters with [escape character](https://api.jquery.com/category/selectors/).

Reply:
I went with urlEncode as it's a very wide range escape solution.

Let me test it out by escaping using that list. Although I think that we still need a special escape for space `" "`.

Reply:
The same with ``"`'`` (double quote, backtick, single quote)

Reply:
> I think that we still need a special escape for space " " (in particular, to prevent problems if id is used in class)

Yes, space is included in this set ` !"#$%&'()*+,./:;<=>?@[\]^`{|}~`

Reply:
I meant that "space", and a bunch of other characters, may need to be a special case of escape character.

`<div id="a name.that:hover">`  is not a problem with ids `$("#a\\ name\.that\\:hover")` returns fine

But I'm very wary of the ramifications of having reserved selectors or separators as valid ids. Does that `.` or `:` have a impact? what if the HTML tag ends in `\"`? does it create a corrupt HTML?

Reply:
From my tests escaping the characters doesn't work, as the id parameter in the `UI`/`moduleServer` is used verbatim, we cannot say that the `id` parameter is `"%cbind%"` and in some cases use "\\%cbind\\%"

Nor can we use `"\\%cbind\\%"` as dataname as that will also need to be escaped when used in JQuery.

**Bottom line:** I believe this would be very hard to support and using urlEncode with `%XX -> __uXX__` replacement is the best solution we can hope to get.